src/posix/platform/resolver.cpp

```diff
@@ -0,0 +1,293 @@
+
+namespace ot {
+namespace Posix {
+
+void Resolver::Init(void)
+{
+    memset(mUpstreamTransaction, 0, sizeof(mUpstreamTransaction));
+
+    LoadDnsServerListFromConf();
+}
+
+void Resolver::LoadDnsServerListFromConf(void)
+{
+    std::string   line;
+    std::ifstream fp;
+
+    fp.open(kResolveConfLocation);
+    if (fp.bad())
+    {
+        otLogCritPlat("Cannot read %s for domain name servers, default to 127.0.0.1", kResolveConfLocation);
+        mUpstreamDnsServerCount = 1;
+        assert(inet_pton(AF_INET, "127.0.0.1", &mUpstreamDnsServerList[0]) == 1);
+        ExitNow();
+    }
+
+    while (std::getline(fp, line) && mUpstreamDnsServerCount < kMaxUpstreamServerCount)
+    {
+        if (line.find(kNameserverItem, 0) == 0)
+        {
+            in_addr_t addr;
+
+            if (inet_pton(AF_INET, &line.c_str()[sizeof(kNameserverItem)], &addr) == 1)
+            {
+                otLogCritPlat("Got nameserver #%d: %s", mUpstreamDnsServerCount,
+                              &line.c_str()[sizeof(kNameserverItem)]);
+                mUpstreamDnsServerList[mUpstreamDnsServerCount] = addr;
+                mUpstreamDnsServerCount++;
+            }
+        }
+    }
+
+exit:
+    return;
+}
+
+void Resolver::Query(otPlatDnsUpstreamQuery *aTxn, const otMessage *aQuery)
+{
+    char        packet[kMaxDnsMessageSize];
+    otError     error  = OT_ERROR_NONE;
+    uint16_t    length = otMessageGetLength(aQuery);
+    sockaddr_in serverAddr;
+
+    Transaction *txn;
+
+    VerifyOrExit(length <= kMaxDnsMessageSize, error = OT_ERROR_NO_BUFS);
+    VerifyOrExit(otMessageRead(aQuery, 0, &packet, sizeof(packet)) == length, error = OT_ERROR_NO_BUFS);
+
+    txn = AllocateTransaction(aTxn);
+    VerifyOrExit(txn != nullptr, error = OT_ERROR_NO_BUFS);
+
+    serverAddr.sin_family = AF_INET;
+    serverAddr.sin_port   = htons(53);
+    for (int i = 0; i < mUpstreamDnsServerCount; i++)
+    {
+        serverAddr.sin_addr.s_addr = mUpstreamDnsServerList[i];
+        VerifyOrExit(sendto(txn->mUdpFd, packet, length, 0, (struct sockaddr *)&serverAddr, sizeof(serverAddr)),
+                     error = OT_ERROR_NO_ROUTE);
+    }
+    otLogInfoPlat("Forwarded DNS query %p to %d server(s).", static_cast<void *>(aTxn), mUpstreamDnsServerCount);
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        otLogCritPlat("Failed to forward DNS query %p to server: %d", static_cast<void *>(aTxn), error);
+    }
+    return;
+}
+
+void Resolver::Cancel(otPlatDnsUpstreamQuery *aTxn)
+{
+    Transaction *txn = GetTransaction(aTxn);
+
+    CloseTransaction(txn);
+}
+
+Resolver::Transaction *Resolver::AllocateTransaction(otPlatDnsUpstreamQuery *aThreadTxn)
+{
+    int          i;
+    int          error     = 0;
+    int          fdOrError = 0;
+    Transaction *ret       = nullptr;
+
+    for (i = 0; i < kMaxUpstreamTransactionCount; i++)
+    {
+        if (mUpstreamTransaction[i].mThreadTxn == nullptr)
+        {
+            ret       = &mUpstreamTransaction[i];
```

Comment:
Set the `ret` after we successfully do `scoket()` call?

This way we don't need the `if (ret != nullptr && error != 0)` block below?